hw/opentitan/ot_spi_device.c

```diff
@@ -1821,26 +1910,30 @@ static MemTxResult ot_spi_device_buf_read_with_attrs(
     (void)attrs;
     uint32_t val32;
 
-    hwaddr last = addr + size - 1u;
+    hwaddr last = (hwaddr)((uint32_t)addr + size - 1u);
 
-    if (last < SPI_SRAM_PAYLOAD_OFFSET + SPI_SRAM_INGRESS_OFFSET) {
+    if (addr < SPI_SRAM_INGRESS_OFFSET) {
         qemu_log_mask(LOG_GUEST_ERROR,
                       "%s: cannot read egress buffer 0x%" HWADDR_PRIx "\n",
                       __func__, addr);
         return MEMTX_DECODE_ERROR;
     }
-    if (last < SPI_SRAM_CMD_OFFSET + SPI_SRAM_INGRESS_OFFSET) {
-        /* payload buffer */
+
+    if ((addr >= SPI_SRAM_PAYLOAD_OFFSET &&
+         last < (SPI_SRAM_PAYLOAD_OFFSET + SPI_SRAM_PAYLOAD_SIZE)) ||
+        (addr >= SPI_SRAM_TPM_WRITE_OFFSET &&
+         last < (SPI_SRAM_TPM_WRITE_OFFSET + SPI_SRAM_TPM_WRITE_SIZE))) {
         val32 = s->sram[addr >> 2u];
-    } else if (last < SPI_SRAM_ADDR_OFFSET + SPI_SRAM_INGRESS_OFFSET) {
+    } else if (addr >= SPI_SRAM_CMD_OFFSET &&
+               last < (SPI_SRAM_CMD_OFFSET + SPI_SRAM_CMD_SIZE)) {
         /* command FIFO */
         val32 = ((const uint32_t *)s->flash.cmd_fifo.data)[addr >> 2u];
-    } else if (last < SPI_SRAM_ADDR_END + SPI_SRAM_INGRESS_OFFSET) {
+    } else if (addr >= SPI_SRAM_ADDR_OFFSET &&
+               last < (SPI_SRAM_ADDR_OFFSET + SPI_SRAM_ADDR_SIZE)) {
         /* address FIFO */
         val32 = s->flash.address_fifo.data[addr >> 2u];
     } else {
-        /* TPM or not used area */
-        qemu_log_mask(LOG_UNIMP, "%s: TPM not supported 0x%" HWADDR_PRIx "\n",
+        qemu_log_mask(LOG_UNIMP, "%s: Area out of bounds 0x%" HWADDR_PRIx "\n",
                       __func__, addr);
```

Comment:
Change this error to also account for misaligned accesses across buffers?

```suggestion
        qemu_log_mask(LOG_UNIMP, "%s: Invalid ingress buffer access to 0x%" HWADDR_PRIx "-0x%" HWADDR_PRIx "\n",
__func__, addr, last);
```